hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,212 @@
+
+			return types, nil
+		},
+	}
+}
+
+// makeStorageTypesVisitor returns a TypeVisitor to do the creation of dedicated storage types
+func makeStorageTypesVisitor(types astmodel.Types) astmodel.TypeVisitor {
+	factory := &StorageTypeFactory{
+		types: types,
+	}
+
+	result := astmodel.MakeTypeVisitor()
+	result.VisitTypeName = factory.visitTypeName
+	result.VisitObjectType = factory.visitObjectType
+	result.VisitArmType = factory.visitArmType
+	//result.VisitResourceType = factory.visitResourceType
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types astmodel.Types
+}
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	visitedName, err := factory.mapTypeName(name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to map name into storage namespace")
+	}
+
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	visitor *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{},
+) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		propertyType, err := visitor.Visit(prop.PropertyType(), oc.forProperty(prop))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			p := factory.makeStorageProperty(prop, propertyType)
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.MakeStorageType(ot), nil
+}
+
+func (factory *StorageTypeFactory) makeStorageProperty(prop *astmodel.PropertyDefinition, propertyType astmodel.Type) *astmodel.PropertyDefinition {
+	p := prop.WithType(propertyType).
+		MakeOptional().
+		WithoutValidation().
+		WithDescription("")
+
+	// If the property is *string we can simplify it to string
+	// This makes other generated code simpler
+	if ot, ok := p.PropertyType().(*astmodel.OptionalType); ok {
+		if ot.Element().Equals(astmodel.StringType) {
+			p = p.WithType(astmodel.StringType)
+		}
+	}
+
+	return p
+}
+
+// mapTypeName maps an existing type name into the right package for the matching storage type
+// Returns the original instance for any type that does not need to be mapped to storage
+func (factory *StorageTypeFactory) mapTypeName(name astmodel.TypeName) (*astmodel.TypeName, error) {
```

Comment:
minor: usage of `*astmodel.TypeName` rather than `astmodel.TypeName` seems unnessecary, as discussed previously `return astmodel.TypeName{}, err` (while icky) is a thing in Go and it seems like it was only promoted to a ptr here to avoid that?

This may be related to one of the comments I made above about why we tolerate `*astmodel.TypeName` when we hadn't previously.

Reply:
Changed return type to `TypeName` instead of `*Typename`

Reply:
I think since you've changed this, you can do away with the bits talked about [here](https://github.com/Azure/k8s-infra/pull/255#discussion_r489813686) as well

Reply:
Done.